src/InsertionsClient.Core/Api/Providers/InsertionApi.cs

```diff
@@ -180,69 +188,73 @@ internal bool TryValidateManifestFile(string manifestFile, out string details)
             return string.IsNullOrWhiteSpace(details);
         }
 
-        internal bool TryExtractManifestAssets(string manifestFile, out List<Asset> assets, out string details)
+        /// <summary>
+        /// Parses the manifest files and extracts the assets into the given collection.
+        /// </summary>
+        /// <param name="manifestFile">Path to the file that will be parsed.</param>
+        /// <param name="assets">The collection that the new assets will be inserted into.</param>
+        /// <param name="details">Details of the encountered issue in case of an error.
+        /// The value is null or empty in the case of success.</param>
+        /// <returns></returns>
+        internal bool TryExtractManifestAssets(string manifestFile, ICollection<Asset> assets, out string details)
         {
             details = string.Empty;
+            Trace.WriteLine($"Loading the manifest file at path {manifestFile}");
 
             try
             {
                 Manifest? buildManifest = DeserializeManifest(manifestFile);
 
                 if (buildManifest == null)
                 {
-                    assets = new List<Asset>();
                     details = "Failed to read/deserialize manifest file";
                     return false;
                 }
 
                 if (!buildManifest.Validate())
                 {
-                    assets = new List<Asset>();
-                    details = $"Validation of de-serialized {InsertionConstants.ManifestFile} file content failed.";
+                    details = $"Validation of de-serialized manifest file content has failed.";
                     return false;
                 }
 
                 if (buildManifest.Builds == null || buildManifest.Builds.Count == 0)
                 {
-                    assets = new List<Asset>();
                     details = $"Manifest file contains no builds.";
                     return false;
                 }
 
-                Trace.WriteLine($"De-serialized {buildManifest.Builds.Count} builds from {InsertionConstants.ManifestFile}.");
+                Trace.WriteLine($"De-serialized {buildManifest.Builds.Count} builds from the manifest file.");
 
-                ConcurrentDictionary<string, Asset> map = new ConcurrentDictionary<string, Asset>();
-                foreach (Build build in buildManifest.Builds.AsParallel())
+                int initialCollectionSize = assets.Count;
+
+                foreach (Build build in buildManifest.Builds)
                 {
-                    foreach (Asset asset in build.Assets.AsParallel())
+                    if (build.Assets == null)
+                    {
+                        Trace.WriteLine("Manifest file contains build with no assets inside. " +
```

Comment:
Not sure if compiler optimizes this. Possible concating of two strings, instead of just string.Format ($)

Reply:
Divided this into two lines for better readability, because otherwise the line was just too long.
I just checked and it appears that the compiler already joins the literals and calls `Concat` for the rest, so we should be fine.

Reply:
Normally yes, but with string interpolation it may be different: https://sharplab.io/#v2:D4AQTAjAsAUCDMACciDCiDetE+UkALIgLIAUAlJtrjSBAAyIDOiAvIgEQCGHiA1IgAkHAHYB7AO6IAliwwARLgBcApgBVpAWxUA6AHKSAvhwDc1HIdiGgA==

Reply:
You are right. It seems to depend on the type of the object you pass. If it is string, `Concat` is used. Otherwise, `Format` is used: https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYGYAE9MGFMDeqmJ2WMALJgLIAUAlIcaSzAgAyYDOmAvJgCIAhgMwBqTABIBAOwD2Ad0xgeBACJDgAUwAqYALZaAdADlFRnXIDKwAE5gZAcwYBfAQG5mJF6hdA===